server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,666 @@
+        }
+
+        @Override
+        public byte readByte() throws IOException {
+            bytesReadTracker.trackPositions(fileOffset + getFilePointer(), 1);
+            return in.readByte();
+        }
+
+        @Override
+        public void readBytes(byte[] b, int offset, int len) throws IOException {
+            bytesReadTracker.trackPositions(fileOffset + getFilePointer(), len);
+            in.readBytes(b, offset, len);
+        }
+    }
+
+    private enum BytesReadTrackingMode {
+        SIZE_ONLY,
+        SIZE_AND_POSITION
+    }
+
+    /**
+     * Tracks the number of bytes that are read in {@link TrackingReadBytesDirectory}. This tracker supports two modes:
+     * - {@link BytesReadTrackingMode#SIZE_ONLY} which is fast but can be inaccurate if the same position is read multiple times
+     * - {@link BytesReadTrackingMode#SIZE_AND_POSITION} which is accurate but slower and uses more memory as it tracks
+     * the positions of each byte read. We need to use this mode if the reader decodes/decompresses the same block/chunk
+     * multiple times or access the same region multiple times.
+     */
+    private static class BytesReadTracker {
+        private static final short PARTITION = 8192;
+        private BytesReadTrackingMode mode = BytesReadTrackingMode.SIZE_ONLY;
+        private long bytesRead;
+        private final Map<Long, CountedBitSet> buckets = new HashMap<>();
+
+        // Keep the last bitset to avoid looking up it in the HashMap as we should access bytes in the same range in adjacent reads.
+        private long lastPositionKey = -1;
+        private CountedBitSet lastBitSet = null;
+
+        void trackPositions(long position, int length) {
+            if (mode == BytesReadTrackingMode.SIZE_ONLY) {
+                assert buckets.isEmpty();
+                bytesRead += length;
+            } else {
+                assert bytesRead == 0;
+                final long endPosition = position + length;
+                for (; position < endPosition; position++) {
+                    long key = position / PARTITION;
+                    if (key != lastPositionKey) {
+                        lastPositionKey = key;
+                        lastBitSet = buckets.computeIfAbsent(lastPositionKey, k -> new CountedBitSet(PARTITION));
+                    }
+                    lastBitSet.set(Math.toIntExact(position % PARTITION));
+                }
+            }
+        }
+
+        void reset(BytesReadTrackingMode mode) {
+            this.mode = mode;
+            this.buckets.clear();
+            this.bytesRead = 0;
+            this.lastPositionKey = -1;
+            this.lastBitSet = null;
+        }
+
+        long getBytesRead() {
+            if (mode == BytesReadTrackingMode.SIZE_AND_POSITION) {
+                assert bytesRead == 0;
+                return buckets.values().stream().mapToLong(CountedBitSet::cardinality).sum();
```

Comment:
I wonder if we should have some heuristics, e.g. in the case when we see small gaps in the middle of long sequences of bytes read, then assume that these small gaps also belong to the field. This could happen e.g. with skip lists of postings, or maybe jump tables of doc values.

Reply:
++, great suggestion. I will work on it.